cardillo/rods/_base.py

```diff
@@ -952,8 +947,8 @@ def g_S_q_T_mu_q(self, t, q, mu):
         coo = CooMatrix((self.nq, self.nq))
         for node in range(self.nnodes_p):
             nodalDOF = self.nodalDOF_p[node]
-            nodalDOF_S = self.nodalDOF_la_S[node]
-            coo[nodalDOF, nodalDOF] = 2 * mu[node] * np.eye(4)
+            # nodalDOF_S = self.nodalDOF_la_S[node]
+            coo[nodalDOF, nodalDOF] = 2 * mu[node] * I4
```

Comment:
This is a diagonal matrix. Can we assemble that more efficiently? Same applies for `g_S_q`. Also `g_S` might utilize some vectorization if we first slice `q` to `P` of shape `n, 4` and subsequently compute something like `g_S = np.inner(p, p, axis=0) - 1`.